### ISSUE TITLE
Missing "?" optional indicator

### ISSUE THREAD

**Issue by codalogic** [arineng/jcrvalidator#7]
The original ABNF allowed ? to indicate an optional object member.  It looks like that's been lost in the Ruby code and you have to do 0*1.  Is that intentional?  Allowing the ? operator seems less noisy.

I would also include the other Kleene operators (bare * and +) as they are the most common repetition cases and people will expect them from regex and so on.

Should:

    p_integer.as(:repetition_min) >> spcCmnt? >> str('*')

be included too (as per float ranges)?

(As an asdie, the integer_r rule should probably have the same layout as float_r, unless .. on it's own is permitted.)

**Comment by anewton1998** [comment]
This was taken out in an effort to reduce redundant ways to express things. But I think you have a point. We should add back "?" and add "+" to be one or more. Additionally, the current code treats "*" as "1*1" instead of zero to infintiy, which is what the current draft says. I should fix the code.

**Comment by codalogic** [comment]
Sounds good.

**Comment by codalogic** [comment]
Should we also allow just a standalone number as the repetition value?  e.g.

my_coords "coords" [ 2 : float ]

**Comment by anewton1998** [comment]
If we can get it to parse, yes. That is s good idea.

**Comment by codalogic** [comment]
I've implemented some enhanced repetition grammar at:

https://github.com/codalogic/jcrvalidator/blob/enhanced_repetitions/lib/jcr/parser.rb (+ some test)

I've realised that in addition to things like the following being valid:
    rule [ *my_type ]

The following is also valid in the current syntax:

    rule [ 0*5my_type ]
and:
    rule2 [ 5my_type ]

That made me think that the grammar should insist on a space after the grammar.  However, the following seems legitimate without a space:
   rule3 [5:int]

and it would be difficult to insist on a space in one context, but not in the other.
So I think I'm happy without insisting on a space, but thought I'd check that you thought the same before making a pull request.
